src/include/storage/delta_record.h

```diff
@@ -0,0 +1,273 @@
+   * @return byte pointer to the attribute. reinterpret_cast and dereference to access the value. if attribute is
+   * nullable and set to null, then return value is nullptr
+   */
+  byte *AccessWithNullCheck(const uint16_t offset) {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    if (!Bitmap().Test(offset)) return nullptr;
+    return reinterpret_cast<byte *>(this) + AttrValueOffsets()[offset];
+  }
+
+  /**
+   * Access a single attribute within the ProjectedRow with a check of the null bitmap first for nullable types
+   * @param offset The 0-indexed element to access in this ProjectedRow
+   * @return byte pointer to the attribute. reinterpret_cast and dereference to access the value. if attribute is
+   * nullable and set to null, then return value is nullptr
+   */
+  const byte *AccessWithNullCheck(const uint16_t offset) const {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    if (!Bitmap().Test(offset)) return nullptr;
+    return reinterpret_cast<const byte *>(this) + AttrValueOffsets()[offset];
+  }
+
+  /**
+   * Access a single attribute within the ProjectedRow without a check of the null bitmap first
+   * @param offset The 0-indexed element to access in this ProjectedRow
+   * @return byte pointer to the attribute. reinterpret_cast and dereference to access the value
+   */
+  byte *AccessForceNotNull(const uint16_t offset) {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    if (!Bitmap().Test(offset)) Bitmap().Flip(offset);
+    return reinterpret_cast<byte *>(this) + AttrValueOffsets()[offset];
+  }
+
+  /**
+   * Set the attribute in the ProjectedRow to be null using the internal bitmap
+   * @param offset The 0-indexed element to access in this ProjectedRow
+   */
+  void SetNull(const uint16_t offset) {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    Bitmap().Set(offset, false);
+  }
+
+  /**
+   * Set the attribute in the ProjectedRow to be not null using the internal bitmap
+   * @param offset The 0-indexed element to access in this ProjectedRow
+   */
+  void SetNotNull(const uint16_t offset) {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    Bitmap().Set(offset, true);
+  }
+
+ private:
+  uint32_t size_;
+  uint16_t num_cols_;
+  byte varlen_contents_[0];
+
+  uint32_t *AttrValueOffsets() { return reinterpret_cast<uint32_t *>(ColumnIds() + num_cols_); }
+
+  const uint32_t *AttrValueOffsets() const { return reinterpret_cast<const uint32_t *>(ColumnIds() + num_cols_); }
+
+  common::RawBitmap &Bitmap() { return *reinterpret_cast<common::RawBitmap *>(AttrValueOffsets() + num_cols_); }
+
+  const common::RawBitmap &Bitmap() const {
+    return *reinterpret_cast<const common::RawBitmap *>(AttrValueOffsets() + num_cols_);
+  }
+};
+
+class DataTable;
+/**
+ * Extension of a ProjectedRow that adds two additional fields: a timestamp and a pointer to the next entry in the
```

Comment:
Nit: Reword this description "timestamp and a pointer" to reflect the class member order. It also adds 4 fields now. We also have a table pointer and a TupleSlot.